### PR TITLE
Changed NP behavior for some edge cases

### PR DESCRIPTION
**Pull request by boris-martin**
## Main changes of this PR

When doing nearest-projection (including the fall-back of linear cell interpolation), we check that the projection we use is closer than the nearest neighbor. If not, we use the NN instead. See #1304 for context.

#### PreferTriangleOverEdge

Consider an edge starting at the origin and moving in the y direction (from (0,0,0) to (0,2,0)) and a triangle living in the plane x = 3. We want to project the point (1,1,1). Before this PR, preCICE tries to map to the triangle, succeeds and thus doesn't check the edge.
Now, we project into the triangle, see that the projection is further than the nearest neighbor vertex (one of the edge), so we decide to reject it. Then we fall back on edges and accept the projection on that edge. **An edge was preferred to a triangle.**

#### AvoidClosestTriangle

Consider two disconnected triangles and a vertex such that only one triangle contains its projection, but it's the furthest of the triangles. Before this PR, we favored interpolation on triangle. Now, we reject this triangle because the projection is further than the nearest neighbor (a vertex from the other triangle). Since the other triangle yields an extrapolation, we reject it, fall-back on edges (there are none) then on vertices. **The far triangle was replaced by the nearest vertex of the other triange**

#### LinearCellInterpolation unit test
See https://github.com/precice/precice/issues/1304: a point fromthe lower-left "AC" region now maps to B, the nearest neighbor.

### 

- Current implementation is a bit suboptimal in 3D: when finding triangle projection, we compare it to NN, and if we fall-back on edges, the NN is recomputed. We could refactor this to gain efficiency at the cost of readability. Let's first discuss the concept.
- Most edge cases happen in volume coupling. If we don't want to change NP for some reason (backward compatibility etc) we could design an "alternative NP" that is only used for the fall-back of LCI. 

Typical example in volume coupling: round domain with different discretizations.
image
The vertex can't be mapped to the nearest edges (it'd be an extrapolation) so it gets mapped to the *opposite* edge.
(In practice, the pentagon would be triangulated and an edge inside it would probably match too)

## Author's checklist

* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I ran `make format` to ensure everything is formatted correctly.
* [ ] I sticked to C++14 features.
* [ ] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

**Comment by fsimonis** [comment]
The problem is not as dramatic as it seems to be. As both meshes discretize the same geometry, the vast majority of points will end up finding the correct nearest primitive.
The primary issue are nearly matching meshes which are shifted outside the domain.

The following image features a geometry (gray) and a discretization (blue). Vertices located inside the red cones are the ones that currently do not provide correct results. As they will not map on the closest edges.


image

**Comment by fsimonis** [comment]
@boris-martin Are you still planning to work on this PR? If not, please unassign and we will take over. :smiley:

**Comment by boris-martin** [comment]
> @boris-martin Are you still planning to work on this PR? If not, please unassign and we will take over. smiley

Not really. Feel free to ask for reviews afterwards though!

File: src/query/Index.cpp [hunk 1]
```diff
@@ -299,16 +299,21 @@ ProjectionMatch Index::findEdgeProjection(const Eigen::VectorXd &location, int n
...
     }
   }
-  // Pick the smallest candidate by distance
+
+  // Could not find edge projection element, fall back to vertex projection
+  if (candidates.empty()) {
+    return closestVertex;
+  }
+
+  // Prefer the closet vertex if it closer than the closest edge
```
Comment (davidscn) [inline src/query/Index.cpp:308]:
```suggestion
  // Prefer the closest vertex if it closer than the closest edge
```

**Review by davidscn — APPROVED** [review]
Looks sane.

**Comment by fsimonis** [comment]
Changelog still missing

**Comment by davidscn** [comment]
Ready to merge, guessing that @boris-martin has no final concerns.

**Comment by boris-martin** [comment]
Alles gut!

**Pull request merged**